### PR TITLE
Added raw key agreement

### PR DESCRIPTION
**Pull request by sbailey-arm**
Signed-off-by: Samuel Bailey <samuel.bailey@arm.com>

### CODE REVIEW

File: protobuf/psa_raw_key_agreement.proto [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ */
+syntax = "proto3";
+
+package psa_raw_key_agreement;
+
+import "psa_algorithm.proto";
+
+message Operation {
+  psa_algorithm.Algorithm.KeyAgreement alg = 1;
+  string private_key = 2;
```
Comment (ionut-arm) [inline protobuf/psa_raw_key_agreement.proto:13]: Maybe call this argument `private_key_name`, to not confuse it with the literal value?

Comment (sbailey-arm) [reply protobuf/psa_raw_key_agreement.proto]: Sounds like a reasonable idea, I've updated the book PR for this too 🙂.

**Comment by hug-dev** [comment]
Is it a single part operation? Does it make sense to be used independantly?

**Comment by sbailey-arm** [comment]
> 
> 
> Is it a single part operation? Does it make sense to be used independantly?

Yes, this one actually is a single part operation 😅 - its a single function call to use (you just need to provide the name of a private key and the public key data (bytes) of a peer key).

**Comment by hug-dev** [comment]
> Yes, this one actually is a single part operation 😅 - its a single function call to use (you just need to provide the name of a private key and the public key data (bytes) of a peer key).

Great! And is it useful as it is, without key derivation? What would be an use-case for it? Sorry for asking too many questions 😛 As the contracts should *never* be modified I think it's ok to be a bit paranoid.

**Comment by ionut-arm** [comment]
> 
> 
> > Yes, this one actually is a single part operation 😅 - its a single function call to use (you just need to provide the name of a private key and the public key data (bytes) of a peer key).
> 
> Great! And is it useful as it is, without key derivation? What would be an use-case for it? Sorry for asking too many questions 😛 As the contracts should _never_ be modified I think it's ok to be a bit paranoid.

It's not recommended in the spec (the recommendation is to use the full key derivation process), but raw key agreement is essentially something like Finite-Field/Elliptic-Curve Diffie-Hellman, where you exchange a public key with the other party, do some calculations using your private key (hence the two arguments) and out comes a random array of bytes that's shared between the two parties. This is the "raw secret", usually passed through some sort of Key Derivation Function to obtain the actual secrets that will be used to encrypt traffic (for example), but if you only want to perform the raw key exchange this operation is needed.

**Pull request merged**